File: algorithms-data-structures/quick-sort.rb
require 'byebug'

def quick_sort arr
  if arr.length <= 1
    return arr
  end

  left = []
  right = []
  pivot = arr.first

  (1..arr.length-1).each do |idx|
    el = arr[idx]
    if el <= pivot
      left.push el
    else
      right.push el
    end
  end

  return quick_sort(left) + [pivot] + quick_sort(right)
end

def quick_sort_in_place arr, start_e, end_e

  if end_e < 0
    end_e = 0
  end
  if start_e > end_e
    start_e = end_e
  end

  # arr of one element, already sorted, so return
  if start_e == end_e
    return
  end

  # sort the arr in place - find index of the pivot - everything to the left of
  # the pivot is less than the pivot, but not necessarily in order - same goes
  # for the right of the pivot, so we need to continue sorting the left and
  # right sides
  pivot_idx = partition arr, start_e, end_e

  quick_sort_in_place arr, 0, pivot_idx-1 # sort the left side of the pivot
  quick_sort_in_place arr, pivot_idx+1, end_e # sort the right side of the pivot

  return arr
end

# sort a given range within an arr, in place, then RETURNS the store_idx, which is the location of the pivot.
def partition arr, start_e, end_e
  pivot = arr[end_e]
  store_idx = start_e

  (start_e..end_e-1).each do |i|
    el = arr[i]
    if el < pivot
      arr[i] = arr[store_idx]
      arr[store_idx] = el
      store_idx += 1
    end
  end

  arr[end_e] = arr[store_idx]
  arr[store_idx] = pivot
  store_idx
end

arr = [3, 2, 1,5, -2]
p quick_sort_in_place arr, 0, arr.length-1

File: algorithms-data-structures/quick-sort.js
function quickSort (arr, start, len) {
  if (len <= 1) {
    return arr;
  }

  var pivotIdx = partition(arr, start, len);
  quickSort(arr, start, pivotIdx - start); // do not include the element at pivotIdx in the next call to quickSort - it's already in the correct position.
  if (pivotIdx + 1 <= arr.length - 1) {
    quickSort(arr, pivotIdx + 1, len - (pivotIdx - start) - 1);
  }
  return arr;
}

function partition (arr, start, len) {
  var pivot = arr[start + len-1]; // let last element always be pivot
  var runnerIdx = start;
  var temp;
  for (var i = start; i < start + len - 1; i++) {
    if (arr[i] < pivot) {
      temp = arr[i];
      arr[i] = arr[runnerIdx];
      arr[runnerIdx] = temp;
      runnerIdx += 1;
    }
  }

  temp = arr[runnerIdx];
  arr[runnerIdx] = pivot;
  arr[start + len - 1] = temp;
  return runnerIdx;
}

a = [ -1, 3, -1, -14, 4, 5, 6, -2, -3, -4, 24 , -234 , -1123 , 234, 45, 2, 43 , 2, 0, 0 ,23, 304, 03, 0];
 // a = [];
console.log(quickSort(a, 0, a.length));
